File: wardend/internal/auth/access.go
package auth

import "net/http"

// The access model Beacon resolves and wardend enforces (ADR-017). Beacon carries the same tables in
// TypeScript (`beacon/lib/access.ts`); TestAccessVectors pins the pair against a shared set of cases.

// InstanceRole is what a principal may do on one instance. Ordered: each includes the previous.
type InstanceRole string

const (
	InstNone     InstanceRole = ""
	InstViewer   InstanceRole = "viewer"
	InstOperator InstanceRole = "operator"
	InstManager  InstanceRole = "manager"
)

var rank = map[InstanceRole]int{InstNone: 0, InstViewer: 1, InstOperator: 2, InstManager: 3}

// AllInstances is the instance id of a grant covering every instance of the node.
const AllInstances = "*"

// Cap is a power that is not tied to a single instance.
type Cap string

const (
	CapSystemUpdate    Cap = "system.update"
	CapJavaManage      Cap = "java.manage"
	CapInstanceCreate  Cap = "instances.create"
	CapInstanceDestroy Cap = "instances.delete"
	CapMembersManage   Cap = "members.manage"
)

// Action is what a request wants to do to an instance.
type Action string

const (
	ActionRead          Action = "read"
	ActionPower         Action = "power"          // start, stop, restart, kill
	ActionConsoleSend   Action = "console.send"   //
	ActionPlayersAction Action = "players.action" // message, kick
	ActionAccessWrite   Action = "access.write"   // whitelist, bans
	ActionOpsWrite      Action = "ops.write"      //
	ActionConfigWrite   Action = "config.write"   // server.properties, raw properties, config files
	ActionPluginsWrite  Action = "plugins.write"  //
	ActionBackupsWrite  Action = "backups.write"  //
	ActionSettingsWrite Action = "settings.write" // instance settings, upgrade, eula, install
)

var needs = map[Action]InstanceRole{
	ActionRead:          InstViewer,
	ActionPower:         InstOperator,
	ActionConsoleSend:   InstOperator,
	ActionPlayersAction: InstOperator,
	ActionAccessWrite:   InstOperator,
	ActionOpsWrite:      InstManager,
	ActionConfigWrite:   InstManager,
	ActionPluginsWrite:  InstManager,
	ActionBackupsWrite:  InstManager,
	ActionSettingsWrite: InstManager,
}

// stronger returns whichever role grants more.
func stronger(a, b InstanceRole) InstanceRole {
	if rank[a] >= rank[b] {
		return a
	}
	return b
}

// HasCap reports whether the principal holds a host or organization capability.
func (p Principal) HasCap(c Cap) bool {
	for _, got := range p.Caps {
		if got == c {
			return true
		}
	}
	return false
}

// RoleOn is the role the principal holds on one instance, blanket grant included.
func (p Principal) RoleOn(instanceID string) InstanceRole {
	return stronger(p.ACLAll, p.ACL[instanceID])
}

// CanSee reports whether the instance exists as far as this principal is concerned.
func (p Principal) CanSee(instanceID string) bool { return p.RoleOn(instanceID) != InstNone }

// Can reports whether the principal may perform the action on the instance.
func (p Principal) Can(instanceID string, a Action) bool {
	return rank[p.RoleOn(instanceID)] >= rank[needs[a]]
}

// RequireCap wraps a handler that needs a host or organization capability.
func RequireCap(c Cap, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		p, ok := FromContext(r.Context())
		if !ok || !p.HasCap(c) {
			writeErr(w, http.StatusForbidden, "forbidden", string(c)+" is required")
			return
		}
		next(w, r)
	}
}

// OnInstance wraps a handler scoped to `{id}`. An instance the caller has no access to answers 404,
// not 403, so the set of instances on the node does not leak.
func OnInstance(a Action, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		p, ok := FromContext(r.Context())
		id := r.PathValue("id")
		if !ok || !p.CanSee(id) {
			writeErr(w, http.StatusNotFound, "instance_not_found", "instance not found")
			return
		}
		if !p.Can(id, a) {
			writeErr(w, http.StatusForbidden, "forbidden", "role "+string(needs[a])+" is required on this instance")
			return
		}
		next(w, r)
	}
}

File: beacon/lib/access.ts
// The access model shared by Beacon's UI and its token minting (ADR-017). wardend carries the same
// tables in Go (`internal/auth/access.go`); the two must stay in sync — `access.test.ts` pins the
// vectors that `TestAccessVectors` checks on the daemon side.

/** Roles a member can hold on one instance. Ordered: each one includes the previous. */
export const INSTANCE_ROLES = ["viewer", "operator", "manager"] as const;
export type InstanceRole = (typeof INSTANCE_ROLES)[number];

/** Roles inside an organization (Better Auth `organization` plugin). */
export const ORG_ROLES = ["owner", "admin", "member"] as const;
export type OrgRole = (typeof ORG_ROLES)[number];

/** Host and organization powers that are not tied to a single instance. */
export const CAPS = ["system.update", "java.manage", "instances.create", "instances.delete", "members.manage"] as const;
export type Cap = (typeof CAPS)[number];

/** Grant covering every instance of a node, stored as an `instanceAccess` row with this id. */
export const ALL_INSTANCES = "*";

export const isInstanceRole = (v: unknown): v is InstanceRole => INSTANCE_ROLES.includes(v as InstanceRole);

const RANK: Record<InstanceRole, number> = { viewer: 1, operator: 2, manager: 3 };

/** True when `have` is at least as strong as `need`. */
export const atLeast = (have: InstanceRole | undefined, need: InstanceRole): boolean =>
  have !== undefined && RANK[have] >= RANK[need];

/** The stronger of two roles; `undefined` counts as "no access". */
export function strongest(a: InstanceRole | undefined, b: InstanceRole | undefined): InstanceRole | undefined {
  if (!a) return b;
  if (!b) return a;
  return RANK[a] >= RANK[b] ? a : b;
}

/** Actions the UI gates on. wardend enforces the same mapping on the routes listed in ADR-017 §5. */
export type InstanceAction =
  | "read"
  | "power" // start, stop, restart, kill
  | "console.send"
  | "players.action" // message, kick
  | "access.write" // whitelist, bans
  | "ops.write"
  | "config.write" // server.properties, raw properties, config files
  | "plugins.write"
  | "backups.write"
  | "settings.write" // instance settings, upgrade, eula, install
  | "voice.listen" // hear the players from the live view (ADR-019)
  | "voice.speak"; // talk to the players from the live view

const NEEDS: Record<InstanceAction, InstanceRole> = {
  read: "viewer",
  power: "operator",
  "console.send": "operator",
  "players.action": "operator",
  "access.write": "operator",
  "ops.write": "manager",
  "config.write": "manager",
  "plugins.write": "manager",
  "backups.write": "manager",
  "settings.write": "manager",
  "voice.listen": "manager",
  "voice.speak": "operator",
};

export const can = (role: InstanceRole | undefined, action: InstanceAction): boolean => atLeast(role, NEEDS[action]);

/** What an organization role grants by itself, before any per-instance grant. */
export function orgGrants(role: OrgRole): { caps: Cap[]; aclAll?: InstanceRole } {
  switch (role) {
    case "owner":
      return { caps: ["instances.create", "instances.delete", "members.manage"], aclAll: "manager" };
    case "admin":
      return { caps: ["members.manage"], aclAll: "manager" };
    default:
      return { caps: [] };
  }
}

/** Everything a signed token says about one user, and what the daemon reads out of it. */
export type Access = {
  /** Host-level and organization-level powers. */
  caps: Cap[];
  /** Role held on every instance of the node, when the user has a blanket grant. */
  aclAll?: InstanceRole;
  /** Role held on specific instances, keyed by instance id. */
  acl: Record<string, InstanceRole>;
};

/** The role a user holds on one instance, blanket grant included. */
export const roleFor = (access: Access, instanceId: string): InstanceRole | undefined =>
  strongest(access.aclAll, access.acl[instanceId]);

export const labelForInstanceRole: Record<InstanceRole, string> = {
  viewer: "Viewer",
  operator: "Operator",
  manager: "Manager",
};

export const describeInstanceRole: Record<InstanceRole, string> = {
  viewer: "Read the console, metrics and players.",
  operator: "Also start, stop and send commands; manage the whitelist and bans.",
  manager: "Also edit configuration, plugins and backups.",
};

export const labelForOrgRole: Record<OrgRole, string> = { owner: "Owner", admin: "Admin", member: "Member" };
